### PR TITLE
Signup login

### PR DESCRIPTION
**Pull request by HumnaAleem**

### CODE REVIEW

File: consumer/src/pages/TeacherPortal/Signin/login.tsx [hunk 1]
```diff
@@ -1,13 +1,123 @@
...
+		if (password) {
+			createLogin(phone_number, password)
+		}
+	}
+
+	if (redirect) {
+		setTimeout(() => {
+			window.location.replace("/teacher")
+		}, 1500)
+	}
```
Comment (Mudassar045) [inline consumer/src/pages/TeacherPortal/Signin/login.tsx:67]: I don't we need to set `delay` here. As I said above you don't need to have this logic for redirection. You can use `<Redirect exact from="/teacher-login" to="/teacher" />` of react-router.
This will be something like:

```
if(auth.token) {
 return <Redirect exact from="/teacher-login" to="/teacher" />
}

return (
 rest of the html elems
)
```

File: consumer/src/pages/TeacherPortal/Signin/login.tsx [hunk 2]
```diff
@@ -1,13 +1,123 @@
...
+			color="primary" >Sign In </Typography>
+		<TextField
+			variant="outlined"
+			label="Phone Number"
+			margin="normal"
+			fullWidth
+			placeholder="e.g. 0300 1110000"
+			type="number"
+			error={phone_number.length < 11 || phone_number.length > 11}
+			onChange={(event) => handleChange("phone_number", event)}
```
Comment (Mudassar045) [inline consumer/src/pages/TeacherPortal/Signin/login.tsx:83]: You don't need to do that way. You can set the `name='phone_number'` as html element attribute and you can get the `name` (key for state object) and `value` from the event directly.

File: consumer/src/pages/TeacherPortal/Signin/register.tsx [hunk 3]
```diff
@@ -1,13 +1,135 @@
...
+import OutlinedInput from '@material-ui/core/OutlinedInput'
+import InputLabel from '@material-ui/core/InputLabel'
+import InputAdornment from '@material-ui/core/InputAdornment'
+
+import '../../SignUp/style.css'
 
 type P = {
+	validation: (number: string, password: string) => void;
+	createAccount: (number: string, password: string, profile: Partial<TeacherProfile>) => void;
+}
```
Comment (Mudassar045) [inline consumer/src/pages/TeacherPortal/Signin/register.tsx:15]: what is `validation: () => void` here ?

File: consumer/src/pages/TeacherPortal/Signin/register.tsx [hunk 4]
```diff
@@ -1,13 +1,135 @@
...
 
+type S = {
+	phone_number: string;
+	password: string;
+	gender: "M" | "F" | undefined;
+	name: string;
```
Comment (Mudassar045) [inline consumer/src/pages/TeacherPortal/Signin/register.tsx:21]: If you want to make gender as undefined you could do this `gender?: "M" | "F"`. It would be better to drive the properties from `TeacherProfile` instead re-writing this way, Also you don't need to have `semicolon` in type or interface across the application code base.

File: consumer/src/pages/TeacherPortal/Signin/register.tsx [hunk 5]
```diff
@@ -1,13 +1,135 @@
...
+		gender: "M",
+		name: "",
+		showPassword: false,
+	})
+
+	const { phone_number, password, gender, name, showPassword } = state
+
+	const handleChange = (type: string, event: any) => {
+		setState({ ...state, [type]: event.target.value });
+	}
```
Comment (Mudassar045) [inline consumer/src/pages/TeacherPortal/Signin/register.tsx:39]: As I said above, you don't need to do that way. You can write
```
	const handle_change = (event: React.ChangeEvent<HTMLInputElement>) => {
		const { name, value } = event.target
		setState({ ...state, [name]: value })
	}
```
Also you don't have to keep the `showPassword` in state. You can make a separate useState().

File: consumer/src/pages/TeacherPortal/Signin/register.tsx [hunk 6]
```diff
@@ -1,13 +1,135 @@
...
+		validation(phone_number, password)
+
+		if (password) {
+			createAccount(phone_number, password, {
+				phone: phone_number,
+				gender: gender,
+				name: name
+			})
+		}
+	}
```
Comment (Mudassar045) [inline consumer/src/pages/TeacherPortal/Signin/register.tsx:52]: Again what is the reason of doing validation like this?

File: consumer/src/pages/TeacherPortal/Signin/register.tsx [hunk 7]
```diff
@@ -1,13 +1,135 @@
...
+
+	const handleClickShowPassword = () => {
+		setState({ ...state, showPassword: !state.showPassword })
+	}
+
+	const handleMouseDownPassword = (event: React.MouseEvent<HTMLButtonElement>) => {
+		event.preventDefault();
+	}
```
Comment (Mudassar045) [inline consumer/src/pages/TeacherPortal/Signin/register.tsx:60]: As said above, we don't need of doing like this.

File: consumer/src/pages/TeacherPortal/Signin/register.tsx [hunk 8]
```diff
@@ -1,13 +1,135 @@
...
+				onChange={(event) => handleChange("gender", event)}
+			>
+				<MenuItem value="M">Male</MenuItem>
+				<MenuItem value="F">Female</MenuItem>
+			</TextField>
+			<Box width="1" style={{ textAlign: "center" }} >
+				<Button
+					style={{ width: "20ch", marginBottom: 20, marginTop: 20, background: "#f05967", color: "white", borderRadius: "32px", fontWeight: "bold", fontSize: "1.25rem" }}
+					variant="contained"
+					onClick={onSave}>
```
Comment (Mudassar045) [inline consumer/src/pages/TeacherPortal/Signin/register.tsx:127]: Can we make `css` generic for these buttons ?

File: consumer/src/pages/TeacherPortal/index.tsx [hunk 9]
```diff
@@ -95,9 +95,13 @@ const TeacherPortal: React.FC<P> = ({ auth, teacher_portal, updateTeacherProfile
...
-		setAssessmentId(assessmentId)
-		setVideoId(videoId)
+		if (auth.token) {
+			setAssessmentModal(true)
+			setAssessmentId(assessmentId)
+			setVideoId(videoId)
+		} else {
+			window.location.replace("/teacher-login")
+		}
 	}
```
Comment (Mudassar045) [inline consumer/src/pages/TeacherPortal/index.tsx:105]: You don't need to that. I already handle auth for and redirection to login on click event. Please revert these changes. I disabled the button with this condition `disabled={!auth.token && auth.user !== AppUserRole.TEACHER}`

**Review by Mudassar045 — CHANGES_REQUESTED** [review]
Overall it's good work but there is a few minor changes and some questions for you. Thanks

File: consumer/src/pages/TeacherPortal/Signin/index.tsx [hunk 10]
```diff
@@ -1,16 +1,26 @@
 import React from 'react'
 import { Container, Avatar } from '@material-ui/core'
+import { connect } from 'react-redux'
+import { teacherLogin, teacherSignup } from 'actions'
+import { TeacherRegister } from './register'
+import { TeacherLogin } from './login'
+import { validation } from '../../../utils/teacherPortal'
```
[the remark this answers is not in the document]
Comment (HumnaAleem) [reply consumer/src/pages/TeacherPortal/Signin/index.tsx]: Ok.

File: consumer/src/pages/TeacherPortal/Signin/login.tsx [hunk 11]
```diff
@@ -1,13 +1,123 @@
...
+		password: "",
+		showPassword: false,
+		redirect: false
+	})
+
+	useEffect(() => {
+		if (auth.token) {
+			setState({ ...state, redirect: true })
+		}
+	}, [auth.token])
```
[the remark this answers is not in the document]
Comment (HumnaAleem) [reply consumer/src/pages/TeacherPortal/Signin/login.tsx]: I improved the logic.

File: consumer/src/pages/TeacherPortal/Signin/login.tsx [hunk 12]
```diff
@@ -1,13 +1,123 @@
...
+
+	const { phone_number, password, showPassword, redirect } = state
+
+	const handleClickShowPassword = () => {
+		setState({ ...state, showPassword: !state.showPassword })
+	}
```
[the remark this answers is not in the document]
Comment (HumnaAleem) [reply consumer/src/pages/TeacherPortal/Signin/login.tsx]: removed action handler and write in this way

File: consumer/src/pages/TeacherPortal/Signin/login.tsx [hunk 13]
```diff
@@ -1,13 +1,123 @@
...
+
+	const handleMouseDownPassword = (event: React.MouseEvent<HTMLButtonElement>) => {
+		event.preventDefault();
+	}
```
[the remark this answers is not in the document]
Comment (HumnaAleem) [reply consumer/src/pages/TeacherPortal/Signin/login.tsx]: removed

File: consumer/src/pages/TeacherPortal/Signin/login.tsx [hunk 14]
```diff
@@ -1,13 +1,123 @@
...
+
+	const handleChange = (type: string, event: any) => {
+		setState({ ...state, [type]: event.target.value });
+	}
```
[the remark this answers is not in the document]
Comment (HumnaAleem) [reply consumer/src/pages/TeacherPortal/Signin/login.tsx]: replaced

File: consumer/src/pages/TeacherPortal/Signin/login.tsx (hunk 2 above)
Comment (HumnaAleem) [reply consumer/src/pages/TeacherPortal/Signin/login.tsx]: done

File: consumer/src/pages/TeacherPortal/Signin/register.tsx (hunk 3 above)
Comment (HumnaAleem) [reply consumer/src/pages/TeacherPortal/Signin/register.tsx:15]: validation function

File: consumer/src/pages/TeacherPortal/Signin/register.tsx (hunk 8 above)
Comment (HumnaAleem) [reply consumer/src/pages/TeacherPortal/Signin/register.tsx]: nope

File: consumer/src/pages/TeacherPortal/index.tsx (hunk 9 above)
Comment (HumnaAleem) [reply consumer/src/pages/TeacherPortal/index.tsx]: ok
